File: packages/python/core/tests/units/test_sandbox.py
"""Tests for NCL-driven sandbox executor Python bindings."""

import sys

import pytest

# Import from the Rust bindings
try:
    from omni_core_rs import (
        ExecutionResult,
        MountConfig,
        NsJailExecutor,
        SandboxConfig,
        SeatbeltExecutor,
        sandbox_detect_platform,
        sandbox_is_nsjail_available,
        sandbox_is_seatbelt_available,
    )

    HAS_SANDBOX_BINDINGS = True
except ImportError:
    HAS_SANDBOX_BINDINGS = False


# Skip all tests if bindings are not available
pytestmark = pytest.mark.skipif(
    not HAS_SANDBOX_BINDINGS, reason="omni_core_rs sandbox bindings not available"
)


class TestPlatformDetection:
    """Test platform detection functions."""

    def test_detect_platform_returns_valid_value(self):
        """Platform should be one of: linux, macos, unknown."""
        platform = sandbox_detect_platform()
        assert platform in ["linux", "macos", "unknown"]

    def test_detect_platform_matches_os(self):
        """Platform should match current OS."""

        platform = sandbox_detect_platform()
        if sys.platform == "linux":
            assert platform == "linux"
        elif sys.platform == "darwin":
            assert platform == "macos"

    def test_is_nsjail_available(self):
        """Check nsjail availability function runs without error."""
        result = sandbox_is_nsjail_available()
        assert isinstance(result, bool)

    def test_is_seatbelt_available(self):
        """Check seatbelt availability function runs without error."""
        result = sandbox_is_seatbelt_available()
        assert isinstance(result, bool)


class TestExecutionResult:
    """Test ExecutionResult class."""

    def test_execution_result_creation(self):
        """Test creating an ExecutionResult."""
        result = ExecutionResult(
            success=True,
            exit_code=0,
            stdout="test output",
            stderr="",
            execution_time_ms=100,
            memory_used_bytes=1024,
            error=None,
        )
        assert result.success is True
        assert result.exit_code == 0
        assert result.stdout == "test output"
        assert result.stderr == ""
        assert result.execution_time_ms == 100
        assert result.memory_used_bytes == 1024
        assert result.error is None

    def test_execution_result_error(self):
        """Test ExecutionResult with error."""
        result = ExecutionResult(
            success=False,
            exit_code=1,
            stdout="",
            stderr="command not found",
            execution_time_ms=50,
            memory_used_bytes=None,
            error="Execution failed",
        )
        assert result.success is False
        assert result.exit_code == 1
        assert result.error == "Execution failed"


class TestSandboxConfig:
    """Test SandboxConfig class."""

    def test_sandbox_config_creation(self):
        """Test creating a SandboxConfig."""
        config = SandboxConfig(
            skill_id="test-skill",
            mode="EXEC",
            hostname="test-container",
            cmd=["/bin/ls", "/tmp"],
            env=["PATH=/usr/bin"],
            mounts=[
                MountConfig(
                    src="/tmp",
                    dst="/tmp",
                    fstype="tmpfs",
                    rw=True,
                )
            ],
            rlimit_as=100_000_000,
            rlimit_cpu=60,
            rlimit_fsize=10_000_000,
            seccomp_mode=2,
            log_level="info",
        )

        assert config.skill_id == "test-skill"
        assert config.mode == "EXEC"
        assert config.hostname == "test-container"
        assert len(config.cmd) == 2
        assert len(config.env) == 1
        assert len(config.mounts) == 1
        assert config.rlimit_as == 100_000_000
        assert config.rlimit_cpu == 60
        assert config.seccomp_mode == 2

    def test_sandbox_config_empty_mounts(self):
        """Test config with empty mounts."""
        config = SandboxConfig(
            skill_id="test",
            mode="EXEC",
            hostname="test",
            cmd=["/bin/echo"],
            env=[],
            mounts=[],
            rlimit_as=0,
            rlimit_cpu=0,
            rlimit_fsize=0,
            seccomp_mode=0,
            log_level="info",
        )
        assert len(config.mounts) == 0


class TestMountConfig:
    """Test MountConfig class."""

    def test_mount_config_read_write(self):
        """Test read-write mount."""
        mount = MountConfig(
            src="/data",
            dst="/app/data",
            fstype="bind",
            rw=True,
        )
        assert mount.src == "/data"
        assert mount.dst == "/app/data"
        assert mount.fstype == "bind"
        assert mount.rw is True

    def test_mount_config_read_only(self):
        """Test read-only mount."""
        mount = MountConfig(
            src="/etc",
            dst="/app/etc",
            fstype="bind",
            rw=False,
        )
        assert mount.rw is False


class TestNsJailExecutor:
    """Test NsJailExecutor class."""

    def test_nsjail_executor_creation(self):
        """Test creating NsJailExecutor with default path."""
        executor = NsJailExecutor(None, 60)
        assert executor.name() == "nsjail"

    def test_nsjail_executor_custom_path(self):
        """Test creating NsJailExecutor with custom path."""
        executor = NsJailExecutor("/custom/path/nsjail", 120)
        assert executor.name() == "nsjail"


class TestSeatbeltExecutor:
    """Test SeatbeltExecutor class."""

    def test_seatbelt_executor_creation(self):
        """Test creating SeatbeltExecutor."""
        executor = SeatbeltExecutor(60)
        assert executor.name() == "seatbelt"

    def test_seatbelt_executor_name(self):
        """Test SeatbeltExecutor name is consistent."""
        executor = SeatbeltExecutor(30)
        assert executor.name() == "seatbelt"


class TestSandboxAvailability:
    """Test sandbox availability checking."""

    def test_sandbox_functions_return_bool(self):
        """All availability functions should return boolean."""
        platform = sandbox_detect_platform()
        assert isinstance(platform, str)

        nsjail = sandbox_is_nsjail_available()
        assert isinstance(nsjail, bool)

        seatbelt = sandbox_is_seatbelt_available()
        assert isinstance(seatbelt, bool)

    def test_at_least_one_platform_detected(self):
        """Should detect either linux or macos."""
        platform = sandbox_detect_platform()
        assert platform in ["linux", "macos", "unknown"]

File: packages/rust/crates/omni-sandbox/tests/test_sandbox.rs
//! Integration tests for `omni-sandbox`.

use std::path::PathBuf;

use omni_sandbox::{
    ExecutionResult, MountConfig, NsJailExecutor, SandboxConfig, SeatbeltExecutor, detect_platform,
    is_nsjail_available, is_seatbelt_available,
};
use tempfile::TempDir;

#[test]
fn test_detect_platform() {
    let platform = detect_platform();
    assert!(
        platform == "linux" || platform == "macos" || platform == "unknown",
        "Platform should be valid, got: {platform}"
    );
}

#[test]
fn test_detect_platform_matches_os() {
    let platform = detect_platform();
    #[cfg(target_os = "linux")]
    assert_eq!(platform, "linux");
    #[cfg(target_os = "macos")]
    assert_eq!(platform, "macos");
}

#[test]
fn test_nsjail_availability() {
    let _ = is_nsjail_available();
}

#[test]
fn test_seatbelt_availability() {
    let _ = is_seatbelt_available();
}

#[test]
fn test_sandbox_config_creation() {
    let config = SandboxConfig {
        skill_id: "test-skill".to_string(),
        mode: "EXEC".to_string(),
        hostname: "test-container".to_string(),
        cmd: vec!["/bin/ls".to_string(), "/tmp".to_string()],
        env: vec!["PATH=/usr/bin".to_string()],
        mounts: vec![MountConfig {
            src: "/tmp".to_string(),
            dst: "/tmp".to_string(),
            fstype: "tmpfs".to_string(),
            rw: true,
        }],
        rlimit_as: 100_000_000,
        rlimit_cpu: 60,
        rlimit_fsize: 10_000_000,
        seccomp_mode: 2,
        log_level: "info".to_string(),
    };

    assert_eq!(config.skill_id, "test-skill");
    assert_eq!(config.mode, "EXEC");
    assert_eq!(config.rlimit_as, 100_000_000);
    assert!(!config.cmd.is_empty());
    assert!(!config.env.is_empty());
    assert!(!config.mounts.is_empty());
}

#[test]
fn test_execution_result_creation() {
    let result = ExecutionResult {
        success: true,
        exit_code: Some(0),
        stdout: "test output".to_string(),
        stderr: String::new(),
        execution_time_ms: 100,
        memory_used_bytes: Some(1024),
        error: None,
    };

    assert!(result.success);
    assert_eq!(result.exit_code, Some(0));
    assert_eq!(result.stdout, "test output");
    assert!(result.error.is_none());
}

#[test]
fn test_execution_result_error() {
    let result = ExecutionResult {
        success: false,
        exit_code: Some(1),
        stdout: String::new(),
        stderr: "command not found".to_string(),
        execution_time_ms: 50,
        memory_used_bytes: None,
        error: Some("Execution failed".to_string()),
    };

    assert!(!result.success);
    assert_eq!(result.exit_code, Some(1));
    assert!(result.error.is_some());
}

#[test]
fn test_mount_config_creation() {
    let mount = MountConfig {
        src: "/data".to_string(),
        dst: "/app/data".to_string(),
        fstype: "bind".to_string(),
        rw: false,
    };

    assert_eq!(mount.src, "/data");
    assert_eq!(mount.dst, "/app/data");
    assert!(!mount.rw);
}

#[test]
fn test_nsjail_executor_creation() {
    let executor = NsJailExecutor::new(None, 60);
    assert_eq!(executor.name(), "nsjail");
}

#[test]
fn test_nsjail_executor_custom_path() {
    let executor = NsJailExecutor::new(Some("/custom/path/nsjail".to_string()), 120);
    assert_eq!(executor.name(), "nsjail");
}

#[test]
fn test_seatbelt_executor_creation() {
    let executor = SeatbeltExecutor::new(60);
    assert_eq!(executor.name(), "seatbelt");
}

#[test]
fn test_seatbelt_executor_name() {
    let executor = SeatbeltExecutor::new(30);
    assert_eq!(executor.name(), "seatbelt");
}

#[test]
fn test_executor_names() {
    let nsjail = NsJailExecutor::new(None, 60);
    let seatbelt = SeatbeltExecutor::new(60);

    assert_eq!(nsjail.name(), "nsjail");
    assert_eq!(seatbelt.name(), "seatbelt");
}

#[test]
fn test_config_fields_accessible() {
    let config = SandboxConfig {
        skill_id: "field-test".to_string(),
        mode: "EXEC".to_string(),
        hostname: "test-host".to_string(),
        cmd: vec!["/bin/ls".to_string()],
        env: vec![],
        mounts: vec![],
        rlimit_as: 1000,
        rlimit_cpu: 10,
        rlimit_fsize: 500,
        seccomp_mode: 0,
        log_level: "info".to_string(),
    };

    let _ = config.skill_id.as_str();
    let _ = config.mode.as_str();
    let _ = config.hostname.as_str();
    let _ = config.cmd.len();
    let _ = config.env.len();
    let _ = config.mounts.len();
    let _ = config.rlimit_as;
    let _ = config.rlimit_cpu;
    let _ = config.rlimit_fsize;
    let _ = config.seccomp_mode;
    let _ = config.log_level.as_str();

    assert_eq!(config.skill_id, "field-test");
}

#[test]
fn test_config_multiple_mounts() {
    let mounts = [
        MountConfig {
            src: "/etc".to_string(),
            dst: "/etc".to_string(),
            fstype: "bind".to_string(),
            rw: false,
        },
        MountConfig {
            src: "/usr/lib".to_string(),
            dst: "/usr/lib".to_string(),
            fstype: "bind".to_string(),
            rw: false,
        },
        MountConfig {
            src: "/tmp".to_string(),
            dst: "/tmp".to_string(),
            fstype: "tmpfs".to_string(),
            rw: true,
        },
    ];

    assert_eq!(mounts.len(), 3);
    assert!(!mounts[0].rw);
    assert!(mounts[2].rw);
}

#[test]
fn test_temp_dir_creation() -> Result<(), Box<dyn std::error::Error>> {
    let temp_dir = TempDir::new()?;
    let path = temp_dir.path();
    assert!(path.exists());
    Ok(())
}

#[test]
fn test_pathbuf_operations() {
    let mut path = PathBuf::new();
    path.push("/usr");
    path.push("bin");
    path.push("ls");

    assert_eq!(path.to_string_lossy(), "/usr/bin/ls");
}
